substrate/validator-sets/pallet/src/lib.rs

```diff
@@ -0,0 +1,206 @@
+    StorageMap<_, Twox64Concat, ValidatorSetInstance, ValidatorSet<T>, OptionQuery>;
+
+  type Key = BoundedVec<u8, MaxKeyLen>;
+
+  /// The key for a given validator set instance coin.
+  #[pallet::storage]
+  #[pallet::getter(fn key)]
+  pub type Keys<T: Config> =
+    StorageMap<_, Twox64Concat, (ValidatorSetInstance, Coin), Key, OptionQuery>;
+
+  /// If an account has voted for a specific key or not. Prevents them from voting multiple times.
+  #[pallet::storage]
+  #[pallet::getter(fn voted)]
+  pub type Voted<T: Config> = StorageMap<_, Blake2_128Concat, (T::AccountId, Key), (), OptionQuery>;
+
+  /// How many times a key has been voted for. Once consensus is reached, the keys will be adopted.
+  #[pallet::storage]
+  #[pallet::getter(fn vote_count)]
+  pub type VoteCount<T: Config> =
+    StorageMap<_, Blake2_128Concat, (ValidatorSetInstance, Coin, Key), u16, ValueQuery>;
+
+  #[pallet::genesis_build]
+  impl<T: Config> GenesisBuild<T> for GenesisConfig<T> {
+    fn build(&self) {
+      // Start at 1 to leave 0 available for a future design where Serai itself has a distinct
+      // validator set for its Tendermint consensus
+      let mut coins = Vec::new();
+      for coin in 0 .. self.coins.0 {
+        coins.push(Coin(coin));
+      }
+
+      let mut participants = Vec::new();
+      for participant in self.participants.clone() {
+        participants.push((participant, self.bond));
+      }
+
+      ValidatorSets::<T>::set(
+        ValidatorSetInstance(Session(0), ValidatorSetIndex(1)),
+        Some(ValidatorSet {
+          bond: self.bond,
+          coins: BoundedVec::try_from(coins).unwrap(),
+          participants: BoundedVec::try_from(participants).unwrap(),
+        }),
+      );
+    }
+  }
+
+  #[pallet::event]
+  #[pallet::generate_deposit(pub(super) fn deposit_event)]
+  pub enum Event<T: Config> {
+    Vote {
+      voter: T::AccountId,
+      instance: ValidatorSetInstance,
+      coin: Coin,
+      key: Key,
+      // Amount of votes the key now has
+      votes: u16,
+    },
+    KeyGen {
+      instance: ValidatorSetInstance,
+      coin: Coin,
+      key: Key,
+    },
+  }
+
+  #[pallet::error]
+  pub enum Error<T> {
+    /// Validator Set doesn't exist.
+    NonExistentValidatorSet,
+    /// Non-validator is voting.
+    NotValidator,
+    /// Validator Set already generated keys.
+    AlreadyGeneratedKeys,
+    /// Vvalidator has already voted for these keys.
+    AlreadyVoted,
+  }
+
+  #[pallet::call]
+  impl<T: Config> Pallet<T> {
+    #[pallet::weight(0)] // TODO
+    pub fn vote(
+      origin: OriginFor<T>,
+      index: ValidatorSetIndex,
+      coin: Coin,
+      key: Key,
+    ) -> DispatchResult {
+      let signer = ensure_signed(origin)?;
+      // TODO: Do we need to check key is within bounds?
+
+      // TODO: Get session
+      let session: Session = Session(0);
+
+      // Confirm a key hasn't been set for this set instance
+      let instance = ValidatorSetInstance(session, index);
```

Comment:
todo && not SetRotation?

Reply:
If this is made a todo!, the static set can't vote on keys. Mind clarifying?

Reply:
Didn't mean a macro, meant making this via comment with context for next steps here. I thought we would need to check if we were generating or rotating keys here but understand now each session will create new instances for each set.